services/knack_location_updater.py

```diff
@@ -0,0 +1,357 @@
+
+
+def main(args):
+    # Parse Arguments
+    app_name = args.app_name
+    container = args.container
+    filter_iso_date_str = format_filter_date(args.date)
+    logger.info(args)
+
+    token = create_login_token()
+
+    # Getting location data from Postgres(t)
+    client_postgrest = utils.postgrest.Postgrest(PGREST_ENDPOINT, token=PGREST_JWT)
+    data = client_postgrest.select(
+        "knack",
+        params={
+            "select": "record, updated_at",
+            "app_id": f"eq.{APP_ID}",
+            "container_id": f"eq.{container}",
+            "updated_at": f"gte.{filter_iso_date_str}",
+        },
+        order_by="record_id",
+    )
+
+    object = CONFIG[app_name][container]["object"]
+    loc_field = CONFIG[app_name][container]["location_field_id"]
+    modified_date_field = CONFIG[app_name][container]["modified_date_field"]
+    update_processed_field = CONFIG[app_name][container]["update_processed_field"]
+    knack_data = [r["record"] for r in data]
+    output_keys = get_output_keys()
+    unmatched_locations = []
+
+    for record in knack_data:
+        if record[loc_field]:  # ignore records that have a null location record
+            point = [
+                record[f"{loc_field}_raw"]["longitude"],
+                record[f"{loc_field}_raw"]["latitude"],
+            ]
+            changed = False
+            for layer in LAYER_CONFIG:
+                params = get_params(layer, point, token)
+                try:
+                    res = point_in_poly(
+                        layer["service_name"], layer["layer_id"], params
+                    )
+                    if not res["features"] and "service_name_secondary" in layer:
+                        # Some layers have a backup secondary layer to check
+                        res = point_in_poly(
+                            layer["service_name_secondary"], layer["layer_id"], params
+                        )
+                    if res.get("error"):
+                        raise Exception(str(res))
+                    if not res["features"]:
+                        # Case 1: we have no features found for our record
+                        record, changed = handle_no_features(changed, layer, record)
+                    else:
+                        # Case 2: We did indeed find features for our record
+                        record, changed = handle_features(changed, layer, record, res)
+                except Exception as e:
+                    logger.info(f"Error handling location ID:{record['id']}")
+                    logger.info(e)
+                    unmatched_locations.append(record)
+                    continue
+        if changed:
+            # Updating a record in Knack
+            record = {key: record[key] for key in output_keys}
+            # Two additional fields for every record:
+            record[update_processed_field] = True
+            record[modified_date_field] = local_timestamp()
```

Comment:
Ok, so we're pulling the knack records from postgrest based on their modified date in postgrest, and those records which have changed get updated in Knack—including a new modified date. So the next time `records_to_postgrest` runs, it will pick up the modified records in Knack and push them to postgrest, updating their modified date again in postgrest. The next time this script runs, the record will be processed again, but no change will be detected. So every changed record gets handled twice. 

Is this too hair-brained? Would it be better for this script to just pull records directly from Knack and update them in Knack? I think this is fine 😵

Reply:
Oh I see, so I guess the way to avoid that would be to just drop the updating of the `modified_date_field`? That way we only touch each record once?

Reply:
This [sr_asset_assign.py](https://github.com/cityofaustin/atd-data-publishing/blob/production/transportation-data-publishing/data_tracker/sr_asset_assign.py) is pretty similar too so it'd be good to have a good pattern set for this one too.

Reply:
This [sr_asset_assign.py](https://github.com/cityofaustin/atd-data-publishing/blob/production/transportation-data-publishing/data_tracker/sr_asset_assign.py) is pretty similar too so it'd be good to have a good pattern set for this one too.

Reply:
Yeah, I lean toward cutting postgres out of this. I think it's fine for maintenance script like this to only interact with the Knack app of concern.